Avoid repeated code on inspektor-gadget and kubectl-gadget integration tests

pawarpranav83 opened the pull request:
## Avoid repeated code on inspektor-gadget and kubectl-gadget integration tests

Fixes #1271 
Removed code redundancy from common tests of `ig/k8s` and `kubectl-gadget` (integration/inspektor-gadget).

## Testing done

Integration tests were passed.
ig: `make test -C ./integration/k8s`
kubectl-gadget: `make integration-tests`

pawarpranav83 commented:
Is this approach alright? Should I refactor other tests similarly? @blanquicet @mauriciovasquezbernal

mauriciovasquezbernal commented:
Hi @pawarpranav83, thanks for working on this. I think your approach is good, however I think we can go one step beyond and completely remove the duplication of the inspektor-gadget (https://github.com/inspektor-gadget/inspektor-gadget/tree/main/integration/inspektor-gadget) and ig on k8s (https://github.com/inspektor-gadget/inspektor-gadget/tree/main/integration/ig/k8s) tests. If you compare the tests in both folder, you'll find that they're almost identical. 

I did some experimentation on https://github.com/inspektor-gadget/inspektor-gadget/commit/f755ff7e9866370155ab2b92553de73761d38765, I think something along the lines can work. It's still missing all the logic to deploy and undeloy inspektor gadget (when using kubectl-gadget) and other components as done in https://github.com/inspektor-gadget/inspektor-gadget/blob/main/integration/inspektor-gadget/integration_test.go.

Please let us know if you have any question.

pawarpranav83 commented:
@mauriciovasquezbernal, the tests are working for both ig and inspektor-gadget, but I am not able to run inspektor-gadget tests locally (checked them with github actions), the spod pod is giving CrashLoopBackOff error, currently looking into it, but any idea how to solve this?

> spod-7mk54                                            1/2     CrashLoopBackOff   5 (7s ago)   3m32s
> 3m41s       Warning   FailedCreatePodSandBox   pod/security-profiles-operator-bd55f7df9-xzpwf             Failed to create pod sandbox: rpc error: code = Unknown desc = [failed to set up sandbox container "3c51360ab7e13bc915132aafae3852dc5431f90a7c42d925e84e24324238836c" network for pod "security-profiles-operator-bd55f7df9-xzpwf": networkPlugin cni failed to set up pod "security-profiles-operator-bd55f7df9-xzpwf_security-profiles-operator" network: failed to list chains: running [/usr/sbin/iptables -t nat -S --wait]: exit status -1: , failed to clean up sandbox container "3c51360ab7e13bc915132aafae3852dc5431f90a7c42d925e84e24324238836c" network for pod "security-profiles-operator-bd55f7df9-xzpwf": networkPlugin cni failed to teardown pod "security-profiles-operator-bd55f7df9-xzpwf_security-profiles-operator" network: running [/usr/sbin/iptables -t nat -D POSTROUTING -s 10.244.0.13 -j CNI-708fc5426727447a8cb93215 -m comment --comment name: "bridge" id: "3c51360ab7e13bc915132aafae3852dc5431f90a7c42d925e84e24324238836c" --wait]: exit status 2: iptables v1.8.4 (nf_tables): Chain 'CNI-708fc5426727447a8cb93215' does not exist...

mauriciovasquezbernal commented:
You can use the following to avoid deploying this pod (it'll make the `TestAuditSeccomp` test to fail, but you can ignore it).

```bash 
$ make INTEGRATION_TESTS_PARAMS="-no-deploy-spo" integration-tests -o kubectl-gadget
```

mauriciovasquezbernal commented on integration/all/integration_test.go:
```diff
@@ -0,0 +1,421 @@
+package main
```
nit: license header.

mauriciovasquezbernal commented on integration/all/integration_test.go:
```diff
@@ -0,0 +1,421 @@
+	}
+}
+
+func testMainInspektorGadget2(m *testing.M) error {
```
Is this used?

mauriciovasquezbernal reviewed (commented):
I don't know if this is ready to review again, leaving two quick comments. 

Other than that, I don't think we need to create a lot of new files. Tests on `integration/ig/k8s/` and `integration/inspektor-gadget/` should be merged in a new `integration/k8s` folder. 

For github actions, we can keep the current workflows without many modifications, only passing the right value of `test-component` in different places around.

pawarpranav83 commented:
> Other than that, I don't think we need to create a lot of new files. Tests on integration/ig/k8s/ and integration/inspektor-gadget/ should be merged in a new integration/k8s folder.

Will rename all to k8s, but I didn't understand what u meant by merging without creating new files.

And to easily see diffs between tests of all and ig/k8s I can just place these files in ig/k8s at least for reviewing, should I do that?

Need to work on `trace_network_test`, but apart from that all tests are passing, can check it using `make test` for ig tests, and `make integration-tests-all` for inspektor-gadget tests.
@mauriciovasquezbernal

mauriciovasquezbernal commented:
>Will rename all to k8s, but I didn't understand what u meant by merging without creating new files 

I mean that this PR should remove the inspektor-gadget (https://github.com/inspektor-gadget/inspektor-gadget/tree/main/integration/inspektor-gadget) and ig on k8s (https://github.com/inspektor-gadget/inspektor-gadget/tree/main/integration/ig/k8s) folders.

pawarpranav83 commented:
Right, I just kept them for my reference, if some tests don't work, will remove them when all changes are made.

mauriciovasquezbernal commented:
> And is it alright if I work on this after 2 weeks? (have my minors from next week)

For sure!

blanquicet commented on integration/all/integration_test.go:
```diff
@@ -0,0 +1,295 @@
+		return -1
+	}
+
+	if *k8sDistro != "" {
+		found := false
+		for _, val := range supportedK8sDistros {
+			if *k8sDistro == val {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			fmt.Fprintf(os.Stderr, "Error: invalid argument '-k8s-distro': %q. Valid values: %s\n",
+				*k8sDistro, strings.Join(supportedK8sDistros, ", "))
+			return -1
+		}
+	}
```
This should be done also for `IgTestComponent`, so I suggest moving it to `initialize` where you're parsing and validating flags in common.

blanquicet commented on integration/all/integration_test.go:
```diff
@@ -0,0 +1,295 @@
+	doNotDeploySPO = flag.Bool("no-deploy-spo", false, "don't deploy the Security Profiles Operator (SPO)")
+
+	k8sDistro = flag.String("k8s-distro", "", "allows to skip tests that are not supported on a given Kubernetes distribution")
+	k8sArch   = flag.String("k8s-arch", "amd64", "allows to skip tests that are not supported on a given CPU architecture")
```
nit: It is unrelated to this PR but we should have a const `K8sArchAMD64` for this value and validate it in `initialize`.

pawarpranav83 replied:
Something like this:
```
K8sArchAMD64 = "amd64"
k8sArch = flag.String("k8s-arch", "", "allows to skip tests that are not supported on a given CPU architecture")

if *k8sArch != K8sArchAMD64 {
		return fmt.Errorf("Error: invalid argument '-k8s-arch': %q.\n", *k8sArch)
}
```
?

blanquicet commented on integration/all/integration_test.go:
```diff
@@ -0,0 +1,295 @@
+	topTimeoutInSeconds              = 10
+)
+
+var (
+	// supportedContainerRuntimes = []string{ContainerRuntimeDocker, ContainerRuntimeContainerd, ContainerRuntimeCRIO}
+	supportedK8sDistros = []string{K8sDistroAKSAzureLinux, K8sDistroAKSUbuntu, K8sDistroARO, K8sDistroMinikubeGH}
+	cleaningUp          = uint32(0)
+
+	// containerRuntime = flag.String("container-runtime", "", "allows to do validation for expected runtime in the tests")
+	image          = flag.String("image", "", "gadget container image")
+	integration    = flag.Bool("integration", false, "run integration tests")
+	dnsTesterImage = flag.String("dnstester-image", "ghcr.io/inspektor-gadget/dnstester:latest", "dnstester container image")
+	testComponent  = flag.String("test-component", "", "run tests for specific component")
+
+	doNotDeployIG  = flag.Bool("no-deploy-ig", false, "don't deploy Inspektor Gadget")
+	doNotDeploySPO = flag.Bool("no-deploy-spo", false, "don't deploy the Security Profiles Operator (SPO)")
+
+	k8sDistro = flag.String("k8s-distro", "", "allows to skip tests that are not supported on a given Kubernetes distribution")
+	k8sArch   = flag.String("k8s-arch", "amd64", "allows to skip tests that are not supported on a given CPU architecture")
+
+	gadgetRepository = flag.String("gadget-repository", "ghcr.io/inspektor-gadget/gadget", "repository where gadget images are stored")
+	gadgetTag        = flag.String("gadget-tag", "latest", "tag used for gadgets's OCI images")
+)
```
nit: Please separate this into three groups, flags specific for ig-k8s, inspektor-gadget and the ones in common.

blanquicet reviewed (commented):
Hi @pawarpranav83. I haven't reviewed the tests yet but only the logic in `integration_tests.go`, and it's looking great already!

pawarpranav83 commented:
> why it has so much rebase conflicts. Can you please shed some light?

This PR is two months old, I think there were several updates to various tests that were removed.
Will update the new tests accordingly.
These tests were removed since they have been combined for ig and inspektor-gadget in `integration/all` to avoid duplication.

pawarpranav83 commented:
@blanquicet this has become too large of a change in a single PR. Should I create several PRs for each test or groups of tests?

pawarpranav83 commented:
@blanquicet can u provide the workflow approval?

pawarpranav83 commented:
/cc @blanquicet

blanquicet commented on integration/all/trace_dns_test.go:
```diff
@@ -44,8 +44,26 @@ func newTraceDnsCmd(t *testing.T, ns string, dnsServerArgs string) *Command {
 	nslookupCmds := []string{
 		fmt.Sprintf("setuidgid 1000:1111 nslookup -type=a fake.test.com. %s", dnsServer),
 		fmt.Sprintf("setuidgid 1000:1111 nslookup -type=aaaa fake.test.com. %s", dnsServer),
-		fmt.Sprintf("setuidgid 1000:1111 nslookup -type=a nodomain.fake.test.com. %s", dnsServer),
```
Why is this gone?

pawarpranav83 replied:
Added later in the switch statement: 
`nslookupCmds = append(nslookupCmds, fmt.Sprintf("setuidgid 1000:1111 nslookup -type=a nodomain.fake.test.com. %s", dnsServer))`

blanquicet commented on integration/all/trace_mount_test.go:
```diff
@@ -52,16 +66,16 @@ func TestTraceMount(t *testing.T) {
 				e.Runtime.ContainerID = ""
 				e.Runtime.ContainerImageDigest = ""
+
+				if DefaultTestComponent == InspektorGadgetTestComponent {
+					e.Data = ""
+					e.FlagsRaw = 0
+				}
```
Please do this for both.

blanquicet commented on integration/all/trace_network_test.go:
```diff
@@ -105,20 +103,38 @@ func TestTraceNetwork(t *testing.T) {
+				expectedEntries[1].Runtime.ContainerName = "nginx-pod"
+				expectedEntries[1].Runtime.RuntimeName = eventtypes.String2RuntimeName(containerRuntime)
+			} else if DefaultTestComponent == InspektorGadgetTestComponent {
+				expectedEntries[0].Event = BuildBaseEventK8s(ns, WithContainerImageName("docker.io/library/busybox:latest", isDockerRuntime))
+				expectedEntries[0].PodIP = testPodIP
+				expectedEntries[0].PodLabels = map[string]string{"run": "test-pod"}
```
```suggestion
				expectedEntries[0].Event = BuildBaseEventK8s(ns,
					WithContainerImageName("docker.io/library/busybox:latest", isDockerRuntime),
					WithPodLabels("test-pod", ns, isCrioRuntime),
				)
				expectedEntries[0].PodIP = testPodIP
```

pawarpranav83 replied:
`WithPodLabels` sets `expectedEntries[0].K8s.PodLabels`, whereas we need to assign to `expectedEntries[0].PodLabels`

blanquicet commented on integration/all/trace_network_test.go:
```diff
@@ -132,9 +148,25 @@ func TestTraceNetwork(t *testing.T) {
-				// Docker can provide different values for ContainerImageName. See `getContainerImageNamefromImage`
-				if isDockerRuntime {
-					e.Runtime.ContainerImageName = ""
+				if DefaultTestComponent == IgTestComponent {
```
Why not `normalizeCommonData`?

blanquicet reviewed (commented):
It's a great job what you're doing with this PR. It allows us to see what we are doing (wrongly) differently between `ig` and `kubectl-gadget`. In fact, half of the comments are to take the opportunity and unify things, not because there's something wrong with what you did 🙂.

If you move all the `integration/inspektor-gadget` tests to `integration/all`, the Makefile/CI changes will be simpler.

blanquicet commented on integration/all/profile_cpu_test.go:
```diff
@@ -27,31 +26,27 @@ func TestProfileCpu(t *testing.T) {
 	profileCPUCmd := &Command{
 		Name: "ProfileCpu",
-		Cmd:  fmt.Sprintf("ig profile cpu -K -o json --runtimes=%s --timeout 10", *containerRuntime),
+		Cmd:  fmt.Sprintf("%s profile cpu -K -o json --timeout 15 %s", DefaultTestComponent, extraArgs),
```
You're right. Let's address it in another PR.

blanquicet commented on integration/all/top_blockio_test.go:
```diff
@@ -79,12 +72,30 @@ func TestTopBlockIO(t *testing.T) {
+
+	switch DefaultTestComponent {
+	case IgTestComponent:
+		extraArgs = "--runtimes=" + containerRuntime + " -m 999"
```
Mmmm [the issue](https://github.com/inspektor-gadget/inspektor-gadget/pull/1408) with `interval == timeout` wasn't related with `max-rows`.

I just remember why we were using `max-rows` in only `ig`. It's because we are not filtering by namespace so other tests running in parallel could interfere:

https://github.com/inspektor-gadget/inspektor-gadget/blob/009a293fb23dbb19af859017e418da2fe4b17fe9/integration/ig/k8s/top_file_test.go#L97-L101

So, @pawarpranav83, please keep adding `m` only for `ig` and add the previous TODO comment. Thanks!

blanquicet commented on integration/all/integration_test.go:
```diff
@@ -0,0 +1,314 @@
+}
+
+func cleanupFunc(cleanupCommands []*Command) {
+	if !atomic.CompareAndSwapUint32(&cleaningUp, 0, 1) {
+		return
+	}
```
This was already part of our code and I think it's working properly. It ensures `cleanupFunc` is executed only once. I think it does the same of `sync.Once` but less readable hahaha. I think we can move to `sync.Once` but in another PR to avoid mixing things. WDYT?

pawarpranav83 commented on integration/all/Makefile:
```diff
@@ -10,18 +10,15 @@ GOPROXY ?= $(shell go env GOPROXY)
 
 .PHONY: build-tests
 build-tests:
-	docker buildx build --load -t ig-tests -f ./../../../Dockerfiles/ig-tests.Dockerfile \
-		--build-arg GOPROXY=$(GOPROXY) ../../../
```
I referred to ig-tests dockerfile (2 months old ig), at that time GOPROXY wasn't used.

blanquicet replied:
Please add it once you rebase.

pawarpranav83 commented on integration/all/integration_test.go:
```diff
@@ -0,0 +1,314 @@
+// Copyright 2019-2022 The Inspektor Gadget authors
```
Right, I did ask @mauriciovasquezbernal about this, and he suggested that we should have those tests in `integration/inspektor-gadget`.
But you are right, it would remove a lot of duplication.

mauriciovasquezbernal replied:
I think we can put all integration tests in the same folder (perhaps in `/integration/`) and skip kubectl-gadget or ig specific as suggested by @blanquicet above.

pawarpranav83 commented on integration/all/snapshot_process_test.go:
```diff
@@ -0,0 +1,104 @@
+				e.ParentPid = 0
+				e.MountNsID = 0
+
+				e.Runtime.ContainerID = ""
+				e.Runtime.ContainerImageDigest = ""
+
+				if DefaultTestComponent == IgTestComponent {
+					// Docker and CRI-O use a custom container name composed, among
+					// other things, by the pod UID. We don't know the pod UID in
+					// advance, so we can't match the exact expected container name.
+					prefixContainerName := "k8s_" + "test-pod" + "_" + "test-pod" + "_" + ns + "_"
+					if (containerRuntime == ContainerRuntimeDocker || containerRuntime == ContainerRuntimeCRIO) &&
+						strings.HasPrefix(e.Runtime.ContainerName, prefixContainerName) {
+						e.Runtime.ContainerName = "test-pod"
+					}
+
+					// Docker can provide different values for ContainerImageName. See `getContainerImageNamefromImage`
+					if isDockerRuntime {
+						e.Runtime.ContainerImageName = ""
+					}
+				} else if DefaultTestComponent == InspektorGadgetTestComponent {
+					// TODO: Verify container runtime and container name
+					e.Runtime.RuntimeName = ""
+					e.Runtime.ContainerName = ""
+				}
```
We are not normalizing `e.K8s.Node = ""`, rather we set it to: `expectedEntry.K8s.Node = nodeName` on line 60.

mauriciovasquezbernal replied:
Perhaps you can call `normalizeCommonData` and right after do `e.K8s.Node = nodeName`?

pawarpranav83 commented on integration/all/integration_test.go:
```diff
@@ -0,0 +1,314 @@
+			strings.HasPrefix(e.Runtime.ContainerName, prefixContainerName) {
+			e.Runtime.ContainerName = "test-pod"
+		}
+		// Docker can provide different values for ContainerImageName. See `getContainerImageNamefromImage`
+		if isDockerRuntime {
+			e.Runtime.ContainerImageName = ""
+		}
```
Receiving an error after this change, `WithContainerImageName` doesn't set it for docker for the `expectedEntry`, whereas we are normalizing the actual output which does contain the `Runtime.ContainerImageName`, even when the runtime is docker.
So this condition is used because we exclude setting `ContainerImageName` for docker runtime.
Error:
```
{"runtime":{"runtimeName":"docker","containerName":"test-pod","containerImageName":"busybox"},"k8s":{"namespace":"test-trace-open-6234904431641082336","podName":"test-pod","containerName":"test-pod"},"type":"normal","uid":1000,"gid":1111,"comm":"cat","fd":3,"ret":3,"flags":["O_RDONLY"],"mode":"----------","path":"/dev/null"}

expected:
{"runtime":{"runtimeName":"docker","containerName":"test-pod"},"k8s":{"namespace":"test-trace-open-6234904431641082336","podName":"test-pod","containerName":"test-pod"},"type":"normal","uid":1000,"gid":1111,"comm":"cat","fd":3,"ret":3,"flags":["O_RDONLY"],"mode":"----------","path":"/dev/null"}
```

blanquicet commented on integration/all/top_file_test.go:
```diff
@@ -80,9 +52,37 @@ func TestTopFile(t *testing.T) {
+		ReadBytes: 0,
+	}
+
+	switch DefaultTestComponent {
+	case IgTestComponent:
+		extraArgs = "--sort -writes,-wbytes -m 999 --runtimes=" + containerRuntime
+		busyboxCmd = "echo foo > bar"
+		expectedEntry.CommonData = BuildCommonData(ns,
+			WithRuntimeMetadata(containerRuntime),
+			WithContainerImageName("docker.io/library/busybox:latest", isDockerRuntime),
+			WithPodLabels("test-pod", ns, isCrioRuntime),
+		)
+		expectedEntry.Filename = "bar"
+	case InspektorGadgetTestComponent:
+		extraArgs = "--sort \"-writes\" -n " + ns
+		busyboxCmd = "echo date >> /tmp/date.txt"
+		expectedEntry.CommonData = BuildCommonDataK8s(ns, WithContainerImageName("docker.io/library/busybox:latest", isDockerRuntime))
+
+		expectedEntry.Filename = "date.txt"
+	}
```
As per [this comment](https://github.com/inspektor-gadget/inspektor-gadget/pull/2450#discussion_r1549775616), please keep `-m` only for `ig` and add the same TODO. Please repeat this for all top gadgets.

mauriciovasquezbernal commented on integration/all/integration_test.go:
```diff
@@ -0,0 +1,312 @@
+	k8sArch        = flag.String("k8s-arch", "amd64", "allows to skip tests that are not supported on a given CPU architecture")
+)
+
+func normalizeCommonData(e *eventtypes.CommonData, ns string, cn string) {
```
I see cn is only passed in integration/all/trace_open_test.go. I think we can remove this argument and handle the special case there (in integration/all/trace_open_test.go).

mauriciovasquezbernal commented on integration/all/integration_test.go:
```diff
@@ -0,0 +1,312 @@
+	k8sArch        = flag.String("k8s-arch", "amd64", "allows to skip tests that are not supported on a given CPU architecture")
+)
+
+func normalizeCommonData(e *eventtypes.CommonData, ns string, cn string) {
```
This can be implemented in integration/helpers.go.

pawarpranav83 replied:
We would need to pass isDockerRuntime and isCrioRuntime separately.

mauriciovasquezbernal replied:
Right, it's fine to leave it there then.

mauriciovasquezbernal commented on integration/all/list_containers_test.go:
```diff
@@ -93,6 +92,10 @@ func newListContainerTestStep(
 }
 
 func TestListContainers(t *testing.T) {
+	if DefaultTestComponent == InspektorGadgetTestComponent {
+		t.Skip("Skip running list-containers with kubectl-gadget")
+	}
+
```
I think it's safer to make this check the other way around:

```go
	if DefaultTestComponent != IgTestComponent {
		t.Skip("Skip running list-containers without ig-gadget")
	}
```

The reason is that we could add more values to DefaultTestComponent. 
Same for other places.

mauriciovasquezbernal commented on integration/all/profile_blockio_test.go:
```diff
@@ -15,19 +15,28 @@
 
+	var extraArgs string
+	switch DefaultTestComponent {
+	case IgTestComponent:
+		extraArgs = ""
```
nit: This one is not needed.

mauriciovasquezbernal commented on integration/all/profile_tcprtt_test.go:
```diff
@@ -65,44 +71,44 @@ func newProfileTCPRTTCmd(flags string, useTimeout bool, node string, unit histog
 }
 
-func RunTestProfileTCPRTT(t *testing.T) {
+func runTestProfileTCPRTT(t *testing.T) {
```
Any reason not to have it directly on `TestProfileTCPRTT`?

mauriciovasquezbernal commented on integration/all/top_file_test.go:
```diff
@@ -62,12 +49,42 @@ func TestTopFile(t *testing.T) {
+		// 	(1) Increase max-rows to 999 (default is 20)
+		// 	(2) Sort by "-writes,-wbytes" (default is "-reads,-writes,-rbytes,-wbytes")
+		extraArgs = fmt.Sprintf("--sort -writes,-wbytes -m %d --runtimes=%s", maxRows, containerRuntime)
+		busyboxCmd = "echo foo > bar"
```
Do we really need different commands for both cases?

mauriciovasquezbernal commented on integration/all/top_tcp_test.go:
```diff
@@ -93,27 +64,67 @@ func TestTopTcp(t *testing.T) {
+
+	switch DefaultTestComponent {
+	case IgTestComponent:
+		// TODO: Filter by namespace to avoid interferences with events from other
+		// tests. In the meanwhile, given that we are generating events by writing
+		// into a file:
```
I think this comment is not relevant for this case. (In this one we aren't writing to files but creating tcp connections).

mauriciovasquezbernal reviewed (commented):
Some minor comments, but it's looking great! I have some ideas to further reduce the duplication, but I think we can go with this and then improve it further.

mauriciovasquezbernal commented:
@pawarpranav83 please also organize the commits:

- Better option: Have a first commit introducing common logic (github actions, things in helpers.go, etc) and then have a commit converting each gadget.
- Acceptable option: Squash everything into a big commit

mauriciovasquezbernal commented on .github/actions/run-integration-tests-all/action.yml:
Do we really need this file? It seems it's almost identical to .github/actions/run-integration-tests/action.yml.

```diff
$ git diff --no-index run-integration-tests run-integration-tests-all
diff --git run-integration-tests/action.yml run-integration-tests-all/action.yml
index b75eee954..9093bfaee 100644
--- run-integration-tests/action.yml
+++ run-integration-tests-all/action.yml
@@ -1,5 +1,5 @@
-name: "Run integration tests"
-description: "Run Inspektor Gadget integration tests"
+name: "Run integration tests all"
+description: "Run Inspektor Gadget integration tests all"
 
 inputs:
   container_repo:
@@ -17,12 +17,6 @@ inputs:
   dnstester_image:
     description: 'The image used for the dnstester.'
     default: 'ghcr.io/inspektor-gadget/dnstester:latest'
-  gadget_repository:
-    description: 'Repository where gadget images are stored.'
-    required: true
-  gadget_tag:
-    description: 'Tag used for gadgets OCI images'
-    required: true
   test_summary_suffix:
     description: 'Suffix of the test summary file'
     required: true
@@ -35,8 +29,8 @@ runs:
       with:
         name: kubectl-gadget-linux-amd64-tar-gz
         path: /home/runner/work/inspektor-gadget/inspektor-gadget/
-    - name: Integration tests
-      id: integration-tests
+    - name: Integration tests all
+      id: integration-tests-all
       shell: bash
       run: |
         tar zxvf /home/runner/work/inspektor-gadget/inspektor-gadget/kubectl-gadget-linux-amd64.tar.gz
@@ -76,9 +70,7 @@ runs:
           CONTAINER_REPO=${{ inputs.container_repo }} \
           IMAGE_TAG=${{ inputs.image_tag }} \
           DNSTESTER_IMAGE=${{ inputs.dnstester_image }} \
-          GADGET_REPOSITORY=${{ inputs.gadget_repository }} \
-          GADGET_TAG=${{ inputs.gadget_tag }} \
-          -o kubectl-gadget integration-tests |& tee integration.log & wait $!
+          -o kubectl-gadget integration-tests-all |& tee integration.log & wait $!
         echo "IntegrationTestsJob: Done"
     - name: Prepare and publish test reports
       if: always()
```

blanquicet replied:
After applying [this suggestion](https://github.com/inspektor-gadget/inspektor-gadget/pull/2450#discussion_r1556343788), the `run-integration-tests-all` action can be avoided.

mauriciovasquezbernal commented on Makefile:
```diff
@@ -287,6 +287,21 @@ integration-tests: kubectl-gadget
+			-k8s-arch $(KUBERNETES_ARCHITECTURE) \
+			-image $(CONTAINER_REPO):$(IMAGE_TAG) \
+			-dnstester-image $(DNSTESTER_IMAGE) \
+			-test-component=inspektor-gadget \
```
I suppose -test-component needs to be configurable too.

blanquicet replied:
Why configurable? It will always be executed only for inspektor-gadget. For ig, users need to use the `integration/k8s/Makefile`.

mauriciovasquezbernal reviewed (commented):
I think a better place to store the new files is `integration/k8s`. "All" is too generic.

blanquicet commented on integration/all/profile_cpu_test.go:
```diff
@@ -60,10 +55,7 @@ func TestProfileCpu(t *testing.T) {
-				if isDockerRuntime {
-					e.Runtime.ContainerImageName = ""
-				}
+				normalizeCommonData(&e.CommonData, ns)
```
It's not manged yet.

pawarpranav83 commented:
> Better option: Have a first commit introducing common logic (GitHub actions, things in helpers.go, etc) and then have a commit converting each gadget.
Sure, after addressing all the comments and finalizing the code, I will organize the commits this way.

pawarpranav83 commented:
@blanquicet @mauriciovasquezbernal Workflow approval and a final review?

mauriciovasquezbernal commented on integration/k8s/Makefile:
```diff
@@ -1,4 +1,7 @@
-include $(shell pwd)/../../../minikube.mk
+# Test commands in this Makefile only covers Inspektor-gadget. 
```
```suggestion
# Test commands in this Makefile only covers ig. 
```

(I know we have a naming mess here)

mauriciovasquezbernal reviewed (approved):
It looks great to me, thanks a lot for handling all the comments and working on this!

I'll merge once the final nits are handled, feel free to ignore the ones you don't consider fair.

blanquicet commented on integration/k8s/integration_test_kubectl_gadget.go:
```diff
@@ -19,12 +19,11 @@ import (
 	"testing"
 
-	. "github.com/inspektor-gadget/inspektor-gadget/integration"
+	"github.com/inspektor-gadget/inspektor-gadget/integration"
```
nit: I'd prefer to keep importing it with `.`, but I let this up to you.

blanquicet commented on integration/k8s/profile_tcprtt_test.go:
```diff
@@ -12,97 +12,52 @@
 	"testing"
 
-	"github.com/inspektor-gadget/inspektor-gadget/integration"
+	. "github.com/inspektor-gadget/inspektor-gadget/integration"
```
dito

blanquicet reviewed (approved):
Some minor comments but LGTM!

mauriciovasquezbernal reviewed (commented):
Could you also please update some references in the documentation:

```bash
git grep integration/ig/k8s
docs/devel/CONTRIBUTING.md:$ make -C integration/ig/k8s test-all
docs/devel/CONTRIBUTING.md:$ make -C integration/ig/k8s CONTAINER_RUNTIME=containerd test
```
